File: scala-java-compat/Companions.java
public class Companions {
  public static void main(String[] args) {
    System.out.println(TraitCompanion$.MODULE$.CompilerConstant());
    System.out.println(TraitCompanion$.MODULE$.value());

    System.out.println(ClassCompanion$.MODULE$.CompilerConstant());
    System.out.println(ClassCompanion$.MODULE$.value());
    System.out.println(ClassCompanion.CompilerConstant());
    System.out.println(ClassCompanion.value());

    System.out.println(AbstractClassCompanion$.MODULE$.CompilerConstant());
    System.out.println(AbstractClassCompanion$.MODULE$.value());
    System.out.println(AbstractClassCompanion.CompilerConstant());
    System.out.println(AbstractClassCompanion.value());

    System.out.println(SadCompanion$.MODULE$.CompilerConstant());
    System.out.println(SadCompanion$.MODULE$.value());
  }
}

File: scala-java-compat/Companions.scala
abstract trait AbstractTraitCompanion {
}
object AbstractTraitCompanion {
  final val CompilerConstant = 1
  val value = 2
}

trait TraitCompanion {
}
object TraitCompanion {
  final val CompilerConstant = 11
  val value = 12
}

class ClassCompanion {
}
object ClassCompanion {
  final val CompilerConstant = 21
  val value = 22
}

abstract class AbstractClassCompanion {
}
object AbstractClassCompanion {
  final val CompilerConstant = 31
  val value = 32
}

object SadCompanion {
  final val CompilerConstant = 41
  val value = 42
}
